hasktorch/src/Torch/NN.hs

```diff
@@ -54,15 +56,31 @@ data ModelRandomness = Deterministic | Stochastic
 -- https://github.com/hasktorch/hasktorch/blob/35e447da733c3430cd4a181c0e1d1b029b68e942/hasktorch/src/Torch/Random.hs#L38
 data G
 
-type family ModelRandomnessR (out :: Type) :: ModelRandomness where
-  ModelRandomnessR ((->) G (_, G)) = 'Stochastic
-  ModelRandomnessR ((->) G _) =
-    TypeError
-      ( Text "Stochastic models taking a 'Generator' "
-          :<>: Text "(i.e., having the form 'forward :: f -> a -> Generator -> _') "
-          :<>: Text "must return the final Generator state in the form "
-          :<>: Text "'forward :: f -> a -> Generator -> (_, Generator)'"
+-- TODO: move to typelevel utils (maybe Torch.Typed.Aux?)
+type family Contains (f :: k) (a :: Type) :: Bool where
+  Contains a a = 'True
+  Contains (f g) a = Contains f a || Contains g a
+  Contains _ _ = 'False
+
+type family CheckStochasticOutType (f :: k) :: ModelRandomness where
+  CheckStochasticOutType (b, G) =
+    If
+      (Not (Contains b G))
+      'Stochastic
+      ( TypeError
+          ( Text "For stochastic models, 'b' must not contain 'Generator' in "
+              :<>: Text "'forward :: f -> a -> Generator -> (b, Generator)'."
+          )
       )
+  CheckStochasticOutType _ =
+    ( TypeError
+        ( Text "Stochastic models must have a forward pass of the form "
+            :<>: Text "'forward :: f -> a -> Generator -> (b, Generator)'."
+        )
+    )
```

Comment:
@tscholak I wonder if you know if there's a way to use `Contains` here as a constraint, like in your original example, instead of a type family returning `Bool`. E.g. something analogous to a pattern guard, like
```haskell
type family CheckStochasticOutType (f :: k) :: ModelRandomness where
  CheckStochasticOutType (b, G) | Not (Contains b G) = -- ...
```
would be really cool (but this doesn't work, and I haven't found anything analogous yet).

Reply:
hi @mcwitt, yeah, guards would be cool, but that doesn't work.

I think you may just need the `Bool` kind:
```
type family Contains (f :: k) (a :: Type) :: Bool where
  Contains a a = 'True
  Contains (f a) a = 'True
  Contains (f g) a = If (Contains f a) 'True (Contains g a)
  Contains _ a = 'False

type family ModelRandomnessR (out :: Type) :: ModelRandomness where
  ModelRandomnessR ((->) G (a, G)) = If (Contains a G) (TypeError ...) 'Stochastic
  ModelRandomnessR a = If (Contains a G) (TypeError ...) 'Deterministic
```

Reply:
btw, did you find `||` and `If` in http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Type-Bool.html or did you make it yourself? 
I think there's an `If` in `Torch.Typed.Aux` that we should probably remove

Reply:
> btw, did you find Not, ||, and If

Yep, these are from `Data.Type.Bool` (I just found out about this, and remember using `If` from `Torch.Typed.Aux` in a previous PR.) I'll make an issue to remove from `Torch.Typed.Aux`.